haskell-src/Concordium/Types/Transactions.hs

```diff
@@ -36,6 +36,56 @@ import Concordium.Types.Updates
 
 -- * Account transactions
 
+-- | Data common to all transaction types.
+--
+--  * @SPEC: <$DOCS/Transactions#transaction-header>
+data TransactionHeader = TransactionHeader
+  { -- | Sender account.
+    thSender :: AccountAddress,
+    -- | Account nonce.
+    thNonce :: !Nonce,
+    -- | Amount of energy dedicated for the execution of this transaction.
+    thEnergyAmount :: !Energy,
+    -- | Size of the payload in bytes.
+    thPayloadSize :: PayloadSize,
+    -- | Absolute expiration time after which transaction will not be executed
+    thExpiry :: TransactionExpiryTime
+  }
+  deriving (Show, Eq)
+
+-- | The size of a serialized transaction header in bytes.
+transactionHeaderSize :: Word64
+transactionHeaderSize =
+  32 -- AccountAddress (FBS 32)
+    + 8 -- Nonce (Word64)
+    + 8 -- Energy (Word64)
+    + 4 -- PayloadSize (Word32)
+    + 8 -- TransactionExpiryTime (Word64)
+
+-- | Get the size of serialized transactions header and payload in bytes.
+getTransactionHeaderPayloadSize :: TransactionHeader -> Word64
+getTransactionHeaderPayloadSize h = fromIntegral (thPayloadSize h) + transactionHeaderSize
+
+$(deriveJSON defaultOptions {fieldLabelModifier = firstLower . drop 2} ''TransactionHeader)
+
+-- * @SPEC: <$DOCS/Transactions#transaction-header-serialization>
+
+instance S.Serialize TransactionHeader where
+  put TransactionHeader {..} =
+    S.put thSender
+      <> S.put thNonce
+      <> S.put thEnergyAmount
+      <> S.put thPayloadSize
+      <> S.put thExpiry
+
+  get = do
+    thSender <- S.get
+    thNonce <- S.get
+    thEnergyAmount <- S.get
+    thPayloadSize <- S.get
+    thExpiry <- S.get
+    return $! TransactionHeader {..}
+
```

Comment:
I do not understand this change. 

This PR is meant to update the toolchains. We cannot at the same time be moving things around. That is a recipe for problems and needless conflicts.

Is this moved because of a staging restriction, or why is it necessary?

Reply:
I am not in favor of unnecessary changes either. Do you remember our discussion about the `deriveJSON` macro not working? It turned out to be a problem with the order of the declarations. I don't know the exact reason but with the `9.0.2` the current order of the declarations has to be changed to get rid of the compile problems.